File: packages/mapping/src/gateLatch.ts
export class GateLatch {
  private openBuffer: string[] = [];
  private closeBuffer: string[] = [];
  private absentStreak = 0;
  private latched = false;

  constructor(
    private gateOn: Set<string>,
    private gateOff: Set<string>,
    private window = 8,
    private openThreshold = 5,
    private closeThreshold = 5,
    private absentHold = 10
  ) {}

  static fromLists(
    gateOn: string[],
    gateOff: string[],
    opts?: Partial<{
      window: number;
      openThreshold: number;
      closeThreshold: number;
      absentHold: number;
    }>
  ): GateLatch {
    return new GateLatch(
      new Set(gateOn),
      new Set(gateOff.filter((g) => g && g !== "None")),
      opts?.window ?? 8,
      opts?.openThreshold ?? 5,
      opts?.closeThreshold ?? 5,
      opts?.absentHold ?? 10
    );
  }

  get isOpen(): boolean {
    return this.latched;
  }

  update(
    liveGesture: string,
    presence: boolean,
    stableGesture?: string | null
  ): boolean {
    if (!presence) {
      this.absentStreak += 1;
      if (this.absentStreak >= this.absentHold) {
        this.openBuffer = [];
        this.closeBuffer = [];
        this.latched = false;
      }
      return this.latched;
    }

    this.absentStreak = 0;
    const live = liveGesture || "None";
    const stable = stableGesture ?? live;

    if (!this.latched) {
      this.push(this.openBuffer, live);
      if (this.countIn(this.openBuffer, this.gateOn) >= this.openThreshold) {
        this.latched = true;
      }
      return this.latched;
    }

    this.push(this.closeBuffer, stable);
    if (
      this.gateOff.size > 0 &&
      this.countIn(this.closeBuffer, this.gateOff) >= this.closeThreshold
    ) {
      this.latched = false;
      this.openBuffer = [];
      this.closeBuffer = [];
    }
    return this.latched;
  }

  private push(buf: string[], v: string): void {
    buf.push(v);
    if (buf.length > this.window) buf.shift();
  }

  private countIn(buf: string[], labels: Set<string>): number {
    return buf.filter((x) => labels.has(x)).length;
  }
}

File: src/mapping/gate_latch.py
"""Porta de nota com latch: abre com palma (live), fecha com punho (estável)."""

from __future__ import annotations

from collections import deque


class GateLatch:
    """Mantém gate aberto entre gestos intermediários e ausências breves da mão."""

    def __init__(
        self,
        gate_on: list[str],
        gate_off: list[str],
        window: int = 8,
        open_threshold: int = 5,
        close_threshold: int = 5,
        absent_hold: int = 10,
    ) -> None:
        self._gate_on = set(gate_on)
        self._gate_off = {g for g in gate_off if g and g != "None"}
        self._open_buffer: deque[str] = deque(maxlen=window)
        self._close_buffer: deque[str] = deque(maxlen=window)
        self._open_threshold = open_threshold
        self._close_threshold = close_threshold
        self._absent_hold = absent_hold
        self._absent_streak = 0
        self._latched = False

    def update(
        self,
        live_gesture: str,
        presence: bool,
        stable_gesture: str | None = None,
    ) -> bool:
        if not presence:
            self._absent_streak += 1
            if self._absent_streak >= self._absent_hold:
                self._open_buffer.clear()
                self._close_buffer.clear()
                self._latched = False
            return self._latched

        self._absent_streak = 0
        live = live_gesture if live_gesture else "None"
        stable = stable_gesture if stable_gesture else live

        if not self._latched:
            self._open_buffer.append(live)
            if self._count_in(self._open_buffer, self._gate_on) >= self._open_threshold:
                self._latched = True
            return self._latched

        self._close_buffer.append(stable)
        if self._gate_off and self._count_in(
            self._close_buffer, self._gate_off
        ) >= self._close_threshold:
            self._latched = False
            self._open_buffer.clear()
            self._close_buffer.clear()
        return self._latched

    @staticmethod
    def _count_in(buf: deque[str], labels: set[str]) -> int:
        return sum(1 for x in buf if x in labels)
